README.md
[![Build Status](https://travis-ci.org/Haravan/haravan_api.svg?branch=master)](https://travis-ci.org/Haravan/haravan_api)
# Haravan API

The Haravan API gem allows Ruby developers to programmatically access the admin section of Haravan stores.

The API is implemented as JSON over HTTP using all four verbs (GET/POST/PUT/DELETE). Each resource, like Order, Product, or Collection, has its own URL and is manipulated in isolation. In other words, we’ve tried to make the API follow the REST principles as much as possible.

## Usage

### Requirements

All API usage happens through Haravan applications, created by either shop owners for their own shops, or by Haravan Partners for use by other shop owners:

* Shop owners can create applications for themselves through their own admin: http://docs.haravan.com/api/tutorials/creating-a-private-app
* Haravan Partners create applications through their admin: http://app.haravan.com/services/partners

For more information and detailed documentation about the API visit http://api.haravan.com

### Installation

To easily install or upgrade to the latest release, use [gem](http://rubygems.org/)

```bash
gem install haravan_api
```

### Getting Started

HaravanAPI uses ActiveResource to communicate with the REST web service. ActiveResource has to be configured with a fully authorized URL of a particular store first. To obtain that URL you can follow these steps:

1. First create a new application in either the partners admin or your store admin. For a private App you'll need the API_KEY and the PASSWORD otherwise you'll need the API_KEY and SHARED_SECRET.

2. For a private App you just need to set the base site url as follows:

   ```ruby
   shop_url = "https://#{API_KEY}:#{PASSWORD}@SHOP_NAME.myharavan.com/admin"
   HaravanAPI::Base.site = shop_url
   ```

   That's it, you're done, skip to step 6 and start using the API!

   For a partner app you will need to supply two parameters to the Session class before you instantiate it:

  ```ruby
  HaravanAPI::Session.setup({:api_key => API_KEY, :secret => SHARED_SECRET})
  ```

3. In order to access a shop's data, apps need an access token from that specific shop. This is a two-stage process. Before interacting with a shop for the first time an app should redirect the user to the following URL:

   ```
   GET https://SHOP_NAME.myharavan.com/admin/oauth/authorize
   ```

   with the following parameters:

   * ``client_id``– Required – The API key for your app
   * ``scope`` – Required – The list of required scopes (explained here: http://docs.haravan.com/api/tutorials/oauth)
   * ``redirect_uri`` – Optional – The URL that the merchant will be sent to once authentication is complete. Defaults to the URL specified in the application settings and must be the same host as that URL.

   We've added the create_permission_url method to make this easier, first instantiate your session object:

   ```ruby
   session = HaravanAPI::Session.new("SHOP_NAME.myharavan.com")
   ```

   Then call:

   ```ruby
   scope = ["write_products"]
   permission_url = session.create_permission_url(scope)
   ```

   or if you want a custom redirect_uri:

   ```ruby
   permission_url = session.create_permission_url(scope, "https://my_redirect_uri.com")
   ```

4. Once authorized, the shop redirects the owner to the return URL of your application with a parameter named 'code'. This is a temporary token that the app can exchange for a permanent access token. Make the following call:

   ```
   POST https://SHOP_NAME.myharavan.com/admin/oauth/access_token
   ```

   with the following parameters:

   * ``client_id`` – Required – The API key for your app
   * ``client_secret`` – Required – The shared secret for your app
   * ``code`` – Required – The token you received in step 3

   and you'll get your permanent access token back in the response.

   There is a method to make the request and get the token for you. Pass
   all the params received from the previous call and the method will verify
   the params, extract the temp code and then request your token:

   ```ruby
   token = session.request_token(params)
   ```

   This method will save the token to the session object and return it. For future sessions simply pass the token in when creating the session object:

   ```ruby
   session = HaravanAPI::Session.new("SHOP_NAME.myharavan.com", token)
   ```

5. The session must be activated before use:

   ```ruby
   HaravanAPI::Base.activate_session(session)
   ```

6. Now you're ready to make authorized API requests to your shop! Data is returned as ActiveResource instances:

   ```ruby
   shop = HaravanAPI::Shop.current

   # Get a specific product
   product = HaravanAPI::Product.find(179761209)

   # Create a new product
   new_product = HaravanAPI::Product.new
   new_product.title = "Burton Custom Freestlye 151"
   new_product.product_type = "Snowboard"
   new_product.vendor = "Burton"
   new_product.save

   # Update a product
   product.handle = "burton-snowboard"
   product.save
   ```

   Alternatively, you can use #temp to initialize a Session and execute a command which also handles temporarily setting ActiveResource::Base.site:

   ```ruby
   products = HaravanAPI::Session.temp("SHOP_NAME.myharavan.com", token) { HaravanAPI::Product.find(:all) }
   ```

7. If you want to work with another shop, you'll first need to clear the session:

   ```ruby
   HaravanAPI::Base.clear_session
   ```


### Console

This package also includes the ``haravan`` executable to make it easy to open up an interactive console to use the API with a shop.

1. Obtain a private API key and password to use with your shop (step 2 in "Getting Started")

2. Use the ``haravan`` script to save the credentials for the shop to quickly log in.

   ```bash
   haravan add yourshopname
   ```

   Follow the prompts for the shop domain, API key and password.

3. Start the console for the connection.

   ```bash
   haravan console
   ```

4. To see the full list of commands, type:

   ```bash
   haravan help
   ```

## Threadsafety

ActiveResource is inherently non-threadsafe, because class variables like ActiveResource::Base.site and ActiveResource::Base.headers are shared between threads. This can cause conflicts when using threaded libraries, like Sidekiq.

We have a forked version of ActiveResource that stores these class variables in threadlocal variables. Using this forked version will allow HaravanAPI to be used in a threaded environment.

To enable threadsafety with HaravanAPI, add the following to your Gemfile. There are various threadsafe tags that you can use, [depending on which version of rails you are using](https://github.com/haravan/activeresource/tags).

```
gem 'activeresource', git: 'git://github.com/Haravan/activeresource', tag: '4.2-threadsafe'
gem 'haravan_api', '>= 3.2.1'
```

## Using Development Version

Download the source code and run:

```bash
rake install
```

## Additional Resources

API Docs: http://docs.haravan.com/api

Ask questions on the forums: http://ecommerce.haravan.com/c/haravan-apis-and-technology

## Copyright

Copyright (c) 2014 "Haravan Inc.". See LICENSE for details.


CHANGELOG
== 3.2.6 (Unreleased)

* Fixed CustomerSavedSearch#customers method to now correctly return only relevant customers

== Version 3.2.5

* More useful error messages for activating nil sessions
* Add tests for commonly deleted objects, and metafield tests, fix naming error in order_risk_test.rb

== Version 3.2.4

* No API changes

== Version 3.2.3

* Added pry to the CLI

== Version 3.2.2

* Temporary fix for the CLI
* Add a specific exception for signature validation failures

== Version 3.2.1

* Added CarrierService resource
* Added optionally using threadsafe ActiveResource (see readme)
* Fixed bug in validate_signature

== Version 3.2.0

* in Session::request_token params is no longer optional, you must pass all the params and the method will now extract the code
* Fixed JSON errors handling (#103)
* Fixed compatibility with Ruby 2.1.x (#83)
* Fixed getting parent ID from nested resources like Variants (#44)
* Cleaned up compatibility with ActiveResource 4.0.x
* Added OrderRisk resource
* Added FulfillmentService resource
* Removed discontinued ProductSearchEngine resource
* Added convenience method Customer#search (#45)

== Version 3.1.8

* Expose `index` and `show` actions of `Location`
* Added create_permission_url and request_token helper methods
* Edited the readme to better describe the getting started procedure

== Version 3.1.7

* Expose `authors` and `tags` action on Article

== Version 3.1.6

* Add LineItem::Property resource

== Version 3.1.5

* Expose `orders` action on Customer

== Version 3.1.3

* Expose `complete` action on Fulfillment

== Version 3.1.2

* Includes port in domain URI (when other than http/80 or https/443)
* Adds access to CustomerSavedSearch
* Adds resources: Order::DefaultAddress, Client::ClientDetails, Announcement
* Allows access to Articles without a blog_id
* Moves encode and as_json overrides to HaravanAPI::Base scope
* Exposes the `order` action in SmartCollection for general use

== Version 3.0.3

* Add a `customers` helper method to the CustomerGroup resource

== Version 3.0.2

* Brevity in require statements

== Version 3.0.1

* Fix saving nested resources in ActiveResource 3.1+

== Version 3.0.0

* Added support for OAuth Authentication
* Removal of support for Legacy Authentication
* Added Cart resource

== Version 2.3.0

* Fix double root bug with ActiveSupport 3.2.0
* Add metafields methods on Customer resource
* Fix prefix_options on assets returned from Asset.find

== Version 2.2.0

* Fix issues with resources that have both direct and namespaced routes
* Added detailed logger to help with debugging ActiveResource
requests/responses
* Add fulfillment#cancel

== Version 2.1.0

* Fix JSON errors handling
* Remove global limit from HaravanAPI::Limits

== Version 2.0.0

* Bump to 2.0.0 as this release breaks Rails 2 compatibility; we're now officially only supporting Rails 3. Rails 2 devs can follow the rails2 tag in this repo to know where we broke off
* Refactored resources into their own source files
* Added API limits functionality
* Patched ActiveResource issue with roots in JSON
* Added pending, cancelled, accepted, and declined convenience methods to HaravanAPI::RecurringApplicationCharge
* HaravanAPI::Session#temp now available as a convenience method to support temporarily switching to other shops when making calls
* Fixes to `haravan console` CLI tool

== Version 1.2.5

* Fix for Article#comments

== Version 1.2.4

* Added Article#comments
* Added Order#cancel
* Added Comment#restore, #not_spam

== Version 1.2.3

* Added Customer, CustomerGroup support

== Version 1.2.2

* Added ScriptTag support

== Version 1.2.1

* Allow abbreviated names for all commands like rails does, e.g. 'haravan c' instead of 'haravan console'
* Fix Variant to support accessing both nested variants with a product prefix as well as top level variants directly
* Add 'grande' to supported product image size variants

== Version 1.2.0

* Command-line interface
* Allow custom params when fetching a single Asset

== Version 1.1.3 (November 4, 2010)

* Add ProductSearchEngines resource

== Version 1.1.2 (October 20, 2010)

* Fix for users of ActiveResource 3.x

== Version 1.1.1 (October 5, 2010)

* Remove hard coded xml formatting in API calls
* Remove jeweler stuff
* Ruby 1.9 encoding fix

== Version 1.1.0 (September 24, 2010)

* Add new Events API for Shop, Order, Product, CustomCollection, SmartCollection, Page, Blog and Article
* Add new 'compact' product image size variant
* Rails 3 fix: attribute_accessors has to be explicitly included since activesupport 3.0.0

== Version 1.0.6

* Add metafields
* Add latest changes from Haravan including asset support, token validation and a common base class

== Version 1.0.0

* extracting HaravanAPI from Haravan into Gem


lib/haravan_api/session.rb

module HaravanAPI

  class ValidationException < StandardError
  end

  class Session
    cattr_accessor :api_key
    cattr_accessor :secret
    cattr_accessor :protocol
    self.protocol = 'https'

    attr_accessor :url, :token, :name

    class << self

      def setup(params)
        params.each { |k,value| send("#{k}=", value) }
      end

      def temp(domain, token, &block)
        session = new(domain, token)
        begin
          original_domain = host_with_port(HaravanAPI::Base.site.to_s)
        rescue URI::InvalidURIError
        end
        original_token   = HaravanAPI::Base.headers['X-Haravan-Access-Token']
        original_session = new(original_domain, original_token)

        begin
          HaravanAPI::Base.activate_session(session)
          yield
        ensure
          HaravanAPI::Base.activate_session(original_session)
        end
      end

      def prepare_url(url)
        return nil if url.blank?
        url.gsub!(/https?:\/\//, '')                            # remove http:// or https://
        url.concat(".myharavan.com") unless url.include?('.')   # extend url to myharavan.com if no host is given
      end

      def validate_signature(params)
        params = params.with_indifferent_access
        return false unless signature = params[:signature]

        sorted_params = params.except(:signature, :action, :controller).collect{|k,v|"#{k}=#{v}"}.sort.join
        Digest::MD5.hexdigest(secret + sorted_params) == signature
      end

      def host_with_port(site)
        parsed_site = URI.parse(site)
        host = parsed_site.host or return
        port = parsed_site.port
        if (protocol == 'http' && port == 80) || (protocol == 'https' && port == 443)
          host
        else
          "#{host}:#{port}"
        end
      end

    end

    def initialize(url, token = nil)
      self.url, self.token = url, token
      self.class.prepare_url(self.url)
    end

    def create_permission_url(scope, redirect_uri = nil)
      params = {:client_id => api_key, :scope => scope.join(',')}
      params[:redirect_uri] = redirect_uri if redirect_uri
      "#{protocol}://#{url}/admin/oauth/authorize?#{parameterize(params)}"
    end

    def request_token(params)
      return token if token

      unless self.class.validate_signature(params) && params[:timestamp].to_i > 24.hours.ago.utc.to_i
        raise HaravanAPI::ValidationException, "Invalid Signature: Possible malicious login"
      end

      code = params['code']

      response = access_token_request(code)

      if response.code == "200"
        token = JSON.parse(response.body)['access_token']
      else
        raise RuntimeError, response.msg
      end
    end

    def shop
      Shop.current
    end

    def site
      "#{protocol}://#{url}/admin"
    end

    def valid?
      url.present? && token.present?
    end

    private
      def parameterize(params)
        URI.escape(params.collect{|k,v| "#{k}=#{v}"}.join('&'))
      end

      def access_token_request(code)
        uri = URI.parse("#{protocol}://#{url}/admin/oauth/access_token")
        https = Net::HTTP.new(uri.host, uri.port)
        https.use_ssl = true
        request = Net::HTTP::Post.new(uri.request_uri)
        request.set_form_data({"client_id" => api_key, "client_secret" => secret, "code" => code})
        https.request(request)
      end
  end
end
